scripts/obtain/install.ps1

```diff
@@ -3,106 +3,219 @@
+
+    return $VersionInfo
+}
+
+# TODO: AzureChannel and Channel should be unified
+function Get-Azure-Channel-From-Channel([string]$Channel) {
+    Say-Invocation $MyInvocation
+
+    switch ($Channel.ToLower()) {
+        { $_ -eq "nightly" } { return "dev" }
+        { $_ -eq "preview" } { return "beta" }
+        { $_ -eq "production" } { throw "Production channel does not exist yet" }
+        default { throw "``$Channel`` is an invalid channel name. Use one of the following: ``nightly``, ``preview``, ``production``" }
+    }
+}
+
+function Get-Specific-Version-From-Version([string]$AzureFeed, [string]$AzureChannel, [string]$CLIArchitecture, [string]$Version) {
+    Say-Invocation $MyInvocation
+
+    switch ($Version.ToLower()) {
+        { $_ -eq "latest" } {
+            $LatestVersionInfo = Get-Latest-Version-Info -AzureFeed $AzureFeed -AzureChannel $AzureChannel -CLIArchitecture $CLIArchitecture
+            return $LatestVersionInfo.Version
         }
+        { $_ -eq "lkg" } { throw "``-Version LKG`` not supported yet." }
+        default { return $Version }
     }
 }
 
-# Set up the install location
-if (!(Test-Path $InstallDir)) {
-    mkdir $InstallDir | Out-Null
+function Construct-Download-Link([string]$AzureFeed, [string]$AzureChannel, [string]$SpecificVersion, [string]$CLIArchitecture) {
+    Say-Invocation $MyInvocation
+
+    $DownloadLink = "$AzureFeed/$AzureChannel/Binaries/$SpecificVersion/dotnet-win-$CLIArchitecture.$SpecificVersion.zip"
+    Say-Verbose "Constructed Download link: $DownloadLink"
+    return $DownloadLink
 }
 
-# De-powershell the path before passing to .NET APIs
-$InstallDir = Convert-Path $InstallDir
+function Get-User-Share-Path() {
+    Say-Invocation $MyInvocation
+
+    $InstallRoot = $env:DOTNET_INSTALL_DIR
+    if (!$InstallRoot) {
+        $InstallRoot = "$env:LocalAppData\Microsoft\.dotnet"
+    }
+    return $InstallRoot
+}
+
+function Resolve-Installation-Path([string]$InstallDir) {
+    Say-Invocation $MyInvocation
+
+    if ($InstallDir -eq "<usershare>") {
+        return Get-User-Share-Path
+    }
+    return $InstallDir
+}
+
+# TODO: check for global.json
+function Get-Installed-Version-Info([string]$InstallRoot) {
+    Say-Invocation $MyInvocation
+
+    $VersionFile = $InstallRoot + $LocalVersionFileRelativePath
+    Say-Verbose "Local version file: $VersionFile"
+    
+    if (Test-Path $VersionFile) {
+        $VersionText = cat $VersionFile
+        Say-Verbose "Local version file text: $VersionText"
+        return Get-Version-Info-From-Version-Text $VersionText
+    }
+
+    Say-Verbose "Local version file not found."
+
+    return $null
+}
 
-say "Downloading $DotNetFileName from $DotNetUrl"
-$resp = Invoke-WebRequest -UseBasicParsing "$DotNetUrl/$DotNetFileName" -OutFile "$InstallDir\$DotNetFileName"
+function Get-Absolute-Path([string]$RelativeOrAbsolutePath) {
+    # Too much spam
+    # Say-Invocation $MyInvocation
```

Comment:
In general will all of these Say-Invocation calls be necessary?

Are they just for our benefit?